ui/src/components/App.js

```diff
@@ -21,65 +18,105 @@ import HelpContainer from '../containers/HelpContainer';
 import Main from './Main';
 import LoadingScreen from '../components/LoadingScreen/LoadingScreen';
 
-async function requireAuth() {
-  await store.dispatch(getLoginInfo());
-  const { login } = store.getState();
+import { store } from '../store';
+import { serverIO } from '../serverIO';
 
-  if (login.loggedIn) {
-    await store.dispatch(startSession(login.user.inControl));
-  }
+import { getLoginInfo } from '../actions/login';
+import { getInitialState, applicationFetched } from '../actions/general';
 
-  if (login.loggedIn && !serverIO.initialized) {
-    serverIO.listen(store);
+export async function requireAuth() {
+  try {
+    store.dispatch(applicationFetched(false));
+    await store.dispatch(getLoginInfo());
+  } catch {
+    store.dispatch(applicationFetched(true));
+    return;
   }
+
+  await store.dispatch(getInitialState());
```

Comment:
Since the `getInitialState` async action is dispatched only once, when the `App` is rendered for the first time, it would make sense to fetch the login info inside that action as well, in parallel with all the other requests.

You can also move `dispatch(applicationFetched(true));` in there and remove `dispatch(applicationFetched(false));`, I think.

Reply:
Hm, that would perhaps be possible but I don't want to fetch the entire app if `fetchLoginInfo` returns 401, or is that just silly of me ?

Reply:
True, true, it's not a bad idea to want to keep them separate, but it does complexify things. If `fetchLoginInfo` fails, then do you dispatch `getInitialState` again when the user logs in? If you don't, I think it's a bug; if you do, then `getInitialState` is perhaps not the right wording.

It's a choice between slowing down initial load, or slowing down logging in. The ideal UX solution is in between: fast initial load, combined with pre-loading while the user enters their crendentials.

In terms of code, fetching everything on initial load is easier since you don't have to worry about dispatching `getInitialState` again. Note that there are probably opportunities for optimising this, for instance, by combining some endpoints to reduce the number of concurrent requests.